libkbfs/disk_block_cache.go

```diff
@@ -1104,6 +1102,121 @@ func (cache *DiskBlockCacheLocal) DoesCacheHaveSpace(
 	}
 }
 
+// Mark updates the metadata of the given block with the tag.
+func (cache *DiskBlockCacheLocal) Mark(
+	ctx context.Context, blockID kbfsblock.ID, tag string) error {
+	cache.lock.Lock()
+	defer cache.lock.Unlock()
+	err := cache.checkCacheLocked("Block(UpdateMetadata)")
+	if err != nil {
+		return err
+	}
+
+	md, err := cache.getMetadataLocked(blockID, false)
+	if err != nil {
+		return NoSuchBlockError{blockID}
+	}
+	md.Tag = tag
+	return cache.updateMetadataLocked(ctx, blockID.Bytes(), md, false)
+}
+
+func (cache *DiskBlockCacheLocal) deleteNextUnmarkedBatchFromTlf(
+	ctx context.Context, tlfID tlf.ID, tag string, startingKey []byte) (
+	nextKey []byte, err error) {
+	cache.lock.Lock()
+	defer cache.lock.Unlock()
+	err = cache.checkCacheLocked("Block(deleteNextUnmarkedBatchFromTlf)")
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	tlfBytes := tlfID.Bytes()
+	rng := &util.Range{
+		Start: startingKey,
+		Limit: append(tlfBytes, cache.maxBlockID...),
+	}
+	iter := cache.tlfDb.NewIterator(rng, nil)
+	defer iter.Release()
+
+	blockIDs := make([]kbfsblock.ID, 0, numUnmarkedBlocksToCheck)
+	for i := 0; i < numUnmarkedBlocksToCheck; i++ {
+		if !iter.Next() {
+			break
+		}
+		key := iter.Key()
+
+		blockIDBytes := key[len(tlfBytes):]
+		blockID, err := kbfsblock.IDFromBytes(blockIDBytes)
+		if err != nil {
+			cache.log.CWarningf(ctx, "Error decoding block ID %x", blockIDBytes)
+			continue
+		}
+		md, err := cache.getMetadataLocked(blockID, false)
+		if err != nil {
+			cache.log.CWarningf(
+				ctx, "No metadata for %s while checking mark", blockID)
+			continue
+		}
+		if md.Tag != tag {
+			blockIDs = append(blockIDs, blockID)
+		}
+	}
+
+	if iter.Next() {
+		nextKey = iter.Key()
+	}
+
+	if len(blockIDs) > 0 {
+		cache.log.CDebugf(ctx, "Deleting %d unmarked blocks (tag=%s) from %s",
+			len(blockIDs), tag, tlfID)
+		_, _, err = cache.deleteLocked(ctx, blockIDs)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return nextKey, nil
+}
+
+// DeleteUnmarked deletes all the blocks without the given tag.
```

Comment:
It might make more sense for this to return a `<-chan error`, allowing either sync or async usage.

Reply:
As it is now, it's mirroring `ClearAllTlfBlocks`, and I'd prefer to keep it consistent.  If the user wants async, they can just use it in a goroutine themselves, pretty easy.  I'd rather not have to manage the goroutines in the cache if it's not needed.

Reply:
OK fair enough. You can leave it as is then.